### PR TITLE
Lua stack -> FFI, fix bugs and inconsistencies

### PR DESCRIPTION
**Pull request by Atcold**
Add some comments for intelligibility. (Still working on it, do not merge yet.)

### CODE REVIEW

File: TripletCriterion.cu [hunk 1]
```diff
@@ -7,21 +7,35 @@
 __global__ void triplet_dist_kernel(const int n, const float norm,
                                     const int nb_batch, const int length,
                                     float* x, float* y) {
+
+  // Unnecessary for loop...
+  // This is EXACTLY equivalent to `if (i < n)` statement
   for (int i = blockIdx.x*blockDim.x+threadIdx.x; i < n; i += blockDim.x*gridDim.x) {
```
Comment (jhjin) [inline TripletCriterion.cu:14]: This is for GPU thread indexing and it is a essential part when the size of tasks exceeds the number of threads although it will not happen in practice.

**Comment by Atcold** [comment]
I had to move from the _Lua_ stack interface to _FFI_ because Torch has changed internals, moving to `TH` and `THCUNN`. Now the code is independent, and talks to its shared object without knowledge of `nn` or other stuff. Still, `updateGradInput` is giving me segmentation fault now.
Moreover, C and CUDA implementations are giving me different output. I'm trying to figure out what's going on...
If you have some time, I would appreciate some feedback. If you don't, it's fine as well. I'll figure it out myself.

File: TripletCriterion.cu (hunk 1 above)
Comment (Atcold) [reply TripletCriterion.cu:14]: Hmm... you set `n = blockDim.x * gridDim.x`, so it's hard to have `i += n`, no?
Moreover, I believe these are logical _grid_, so _block_s and _thread_s do not have a 1-1 hardware correspondence. I think that _block_s are made of nearby cores, so that the _thread_s can easily share data.

**Comment by Atcold** [comment]
Hey, @jhjin, your code is now bug fixed and consistent. Both _CPU_ and _GPU_ versions output the same result. `gradInput` is now of correct dimensionality, and closes #3. Feel free to review my changes and merge them, if you approve.

File: TripletCriterion.cu (hunk 1 above)
Comment (jhjin) [reply TripletCriterion.cu:14]: Yes, it is hard to have the increment of `i` but good to consider the edge case. Nothing is related to shared memory here. Could you remove the comment? unnecessary for loop is unnecessary.

**Comment by jhjin** [comment]
Appreciate for your nice work. How's training? Does it converge well?

File: TripletCriterion.cu (hunk 1 above)
Comment (Atcold) [reply TripletCriterion.cu:14]: I don't think it's incorrect. `i` will never get an increment, since the increment is equal to `n` and you test for `i < n`. Am I getting this wrong? I'm saying there is no edge case here at all.

**Comment by Atcold** [comment]
**Left**: using random samples.
**Right**: using hard _positive_ and semi-hard _negative_.

screenshot 2016-09-16 17 48 58

File: TripletCriterion.cu (hunk 1 above)
Comment (jhjin) [reply TripletCriterion.cu:14]: it fails `n` is big enough

Comment (Atcold) [reply TripletCriterion.cu:14]: OK, I'm removing it because of this article -> https://devblogs.nvidia.com/parallelforall/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/
Next time, share your wisdom with me 😉

**Comment by jhjin** [comment]
Thanks!
